File: src/schemas/interaction.ts
import { z } from "zod";

export const InteractionTypeSchema = z.enum([
  "scout",
  "application",
  "interview",
]);
export type InteractionType = z.infer<typeof InteractionTypeSchema>;

export const InteractionStatusSchema = z.enum([
  "pending",
  "rejected",
  "interviewing",
  "hired",
]);
export type InteractionStatus = z.infer<typeof InteractionStatusSchema>;

export const ChatMessageSchema = z.object({
  role: z.enum(["company", "agent", "system"]),
  content: z.string(),
  timestamp: z.string().datetime(),
});
export type ChatMessage = z.infer<typeof ChatMessageSchema>;

export const TestResultSchema = z.object({
  passed: z.boolean(),
  score: z.number().min(0).max(100).nullable(),
  response_time_ms: z.number().nonnegative().nullable(),
  error_rate: z.number().min(0).max(100).nullable(),
  details: z.record(z.string(), z.unknown()).nullable(),
  executed_at: z.string().datetime(),
});
export type TestResult = z.infer<typeof TestResultSchema>;

export const InteractionSchema = z.object({
  id: z.string().uuid(),
  agent_id: z.string().uuid(),
  job_id: z.string().uuid(),
  type: InteractionTypeSchema,
  status: InteractionStatusSchema,
  chat_log: z.array(ChatMessageSchema),
  test_result: TestResultSchema.nullable(),
  created_at: z.string().datetime(),
});
export type Interaction = z.infer<typeof InteractionSchema>;

export const CreateInteractionSchema = InteractionSchema.omit({
  id: true,
  chat_log: true,
  test_result: true,
  created_at: true,
});
export type CreateInteractionInput = z.infer<typeof CreateInteractionSchema>;

File: backend/app/models/interaction.py
from pydantic import BaseModel, Field
from typing import Optional
from enum import Enum
from datetime import datetime
import uuid


class InteractionType(str, Enum):
    scout = "scout"          # 企業 → エージェント
    application = "application"  # エージェント → 企業
    interview = "interview"  # 面接（Phase 2）


class InteractionStatus(str, Enum):
    pending = "pending"
    rejected = "rejected"
    interviewing = "interviewing"
    hired = "hired"


class ChatMessage(BaseModel):
    role: str = Field(pattern="^(company|agent|system)$")
    content: str = Field(min_length=1, max_length=10000)
    timestamp: datetime


class TestResult(BaseModel):
    passed: bool
    score: Optional[float] = Field(None, ge=0, le=100)
    response_time_ms: Optional[float] = Field(None, ge=0)
    error_rate: Optional[float] = Field(None, ge=0, le=100)
    details: Optional[dict] = None
    executed_at: datetime


class InteractionCreate(BaseModel):
    agent_id: uuid.UUID
    job_id: uuid.UUID
    type: InteractionType
    message: Optional[str] = Field(None, max_length=5000)


class InteractionStatusUpdate(BaseModel):
    status: InteractionStatus


class ChatMessageCreate(BaseModel):
    content: str = Field(min_length=1, max_length=10000)


class Interaction(BaseModel):
    id: uuid.UUID
    agent_id: uuid.UUID
    job_id: uuid.UUID
    type: InteractionType
    status: InteractionStatus
    chat_log: list[ChatMessage]
    test_result: Optional[TestResult] = None
    created_at: datetime

    model_config = {"from_attributes": True}
